index.js

```diff
@@ -38,9 +39,16 @@ function getConfigs() {
   };
 }
 
+const ThrottledOctokit = Octokit.plugin(throttling);
+
 async function run() {
   const configs = getConfigs();
-  const octokit = new Octokit();
+  const octokit = new ThrottledOctokit({
+    throttle: {
+      onRateLimit: () => {},
+      onAbuseLimit: () => {},
```

Comment:
So setting up the plugin is as simple as it seems to be. Now the question is what we'd like to do/how we'd like to act in response to exceeding the rate limits. There are two types of them:
- https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-rate-limits
   - it's hit when the requests exceed the 5000 requests per hour quote
- https://developer.github.com/v3/guides/best-practices-for-integrators/#dealing-with-abuse-rate-limits
   - it's hit when the requests are burst(i.e. sent in parallel) or sent without authentication(refer to their guidelines)

Should the action retry the request(s) if either of the above occurs? Wouldn't that significantly slow down the runtime?

Reply:
Well, my hope for this plugin is to act as `bottleneck` does so we can actually avoid these scenarios. Is that not the case?

Of course, it's still possible we run into them (e.g. 2 parallel jobs) in which case we should raise and fail. But I assume that'd already happen just by having the request fail so unless my previous point applies I don't think we need this plugin.

Reply:
I don't really get this. You're right in that the package uses `Bottleneck` to throttle the requests:

https://github.com/octokit/plugin-throttling.js/blob/master/src/index.ts#L20

> Is that not the case?

The package implements all the best-practices to prevent hitting the rate limits. I could also corroborate this fact locally, i.e. the requests weren't burst, they were slowed down.

> I don't think we need this plugin.

See https://github.com/c-hive/poe-sniper/pull/367. If the requests are *not* throttled then it'll lead to abusing the rate limit numerous times.

Reply:
> Well, my hope for this plugin is to act as bottleneck does so we can actually avoid these scenarios.

Yea, that'd be my expectation as well but my questions were all about how we should act in response to exceeding the limits.

Reply:
As discussed over the phone:
- let's retry if it's simple to do so
- retry everything if possible(rate limits and abuse rate limits)
- log meaningful messages, investigate if it's possible to set a retry count, if so and if it's reached indicate the failure
- investigate if there's an option to set exponential backoff
- advise setting `timeout-minutes` when using this action(i.e. update `README.md`)

Reply:
> let's retry if it's simple to do so

It's as simple as returning with true.

> investigate if it's possible to set a retry count

There's no such option. We can make it so that the request is retried after a certain amount of seconds(it's an out-of-the-box feature, see above, but note `retryAfter` could be as high as ~3600 seconds in the case of the 5000 requests per hour quota). This way, as you pointed out, we should encourage users to set the `timeout-minutes` option to avoid "neverending" jobs.

> investigate if there's an option to set exponential backoff

No, there isn't.

Reply:
We should set the reties to be reasonably quick, let's say once every 30s. No need for a configuration right now but please name it properly in the code.

Reply:
Let's leave it to retry out of the box. 👍

Reply:
> > let's retry if it's simple to do so
> 
> It's as simple as returning with true.

It would be great to make this explicit, i.e.

```js
retry = true;
// ...
onRateLimit {
  // ...
  return retry;
}
```

Reply:
Looking at the plugin's source-code for a whole, my concern is that it's still unclear to me if the callbacks can be triggered with non-zero `retryCount` value. My gut feeling is that the answer is yes since that's what the conditional check is supposed to do.

Given that it's cumbersome to reproduce rate limit errors(I'm not even sure if there's a way to exceed the rate limits locally, e.g. sending requests with an unauthenticated client would make sense since this means way fewer requests per hour as opposed to authenticated requests but it's denied by `@ockokit/rest`), how about leaving it as-is and testing against sniper?

Reply:
Ah, wait, I can test it against my repo I'm using for local testing.

Reply:
https://github.com/c-hive/poe-sniper/runs/532989083

Cool, now it didn't hit the rate limit. :)